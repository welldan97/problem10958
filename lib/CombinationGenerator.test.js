const CombinationGenerator = require('./CombinationGenerator');
const { allNumbersByOneGenerator } = require('../specifics/numberGenerators');
const { add, subtract } = require('../specifics/operations');

const present = ({ value }) => ({
  value: {
    ...value,
    operations: value.operations.map(o => o.name),
  },
});

describe('CombinationGenerator', () => {
  describe('without batches', () => {
    it('generates by one', () => {
      const operations = [add, subtract];
      const numberGenerator = allNumbersByOneGenerator(3);
      const generator = new CombinationGenerator(numberGenerator, operations);

      expect(present(generator)).toMatchSnapshot();
      expect(generator.done).toEqual(false);
      generator.next();
      expect(present(generator)).toMatchSnapshot();
      expect(generator.done).toEqual(false);
      generator.next();
      expect(present(generator)).toMatchSnapshot();
      expect(generator.done).toEqual(false);
      generator.next();
      expect(present(generator)).toMatchSnapshot();
      expect(generator.done).toEqual(false);
      generator.next();
      expect(present(generator)).toMatchSnapshot();
      expect(generator.done).toEqual(false);
      generator.next();
      expect(present(generator)).toMatchSnapshot();
      expect(generator.done).toEqual(false);
      generator.next();
      expect(present(generator)).toMatchSnapshot();
      expect(generator.done).toEqual(false);
      generator.next();
      expect(present(generator)).toMatchSnapshot();
      expect(generator.done).toEqual(false);
      generator.next();
      expect(present(generator)).toMatchSnapshot();
      expect(generator.done).toEqual(false);
      generator.next();
      expect(present(generator)).toMatchSnapshot();
      expect(generator.done).toEqual(false);
      generator.next();
      expect(present(generator)).toMatchSnapshot();
      expect(generator.done).toEqual(true);
    });
  });

  describe('in batches', () => {
    it('generates by one for given batch from beginning', () => {
      const operations = [add, subtract];
      const generator = new CombinationGenerator(
        allNumbersByOneGenerator(3),
        operations,
        {
          batchNumber: 0,
          batchLength: 3,
        },
      );

      expect(present(generator)).toMatchSnapshot();
      expect(generator.done).toEqual(false);
      generator.next();
      expect(present(generator)).toMatchSnapshot();
      expect(generator.done).toEqual(false);
      generator.next();
      expect(present(generator)).toMatchSnapshot();
      expect(generator.done).toEqual(true);
    });

    it('generates by one for given batch from middle 1', () => {
      const operations = [add, subtract];
      const generator = new CombinationGenerator(
        allNumbersByOneGenerator(3),
        operations,
        {
          batchNumber: 1,
          batchLength: 3,
        },
      );

      expect(present(generator)).toMatchSnapshot();
      expect(generator.done).toEqual(false);
      generator.next();
      expect(present(generator)).toMatchSnapshot();
      expect(generator.done).toEqual(false);
      generator.next();
      expect(present(generator)).toMatchSnapshot();
      expect(generator.done).toEqual(true);
    });

    it('generates by one for given batch from middle 2', () => {
      const operations = [add, subtract];
      const generator = new CombinationGenerator(
        allNumbersByOneGenerator(3),
        operations,
        {
          batchNumber: 2,
          batchLength: 3,
        },
      );

      expect(present(generator)).toMatchSnapshot();
      expect(generator.done).toEqual(false);
      generator.next();
      expect(present(generator)).toMatchSnapshot();
      expect(generator.done).toEqual(false);
      generator.next();
      expect(present(generator)).toMatchSnapshot();
      expect(generator.done).toEqual(true);
    });

    it('generates by one for given batch which starts in the middle of number iteration and ends in the middle of the next one', () => {
      const operations = [add, subtract];
      const generator = new CombinationGenerator(
        allNumbersByOneGenerator(4),
        operations,
        {
          batchNumber: 2,
          batchLength: 4,
        },
      );

      expect(present(generator)).toMatchSnapshot();
      expect(generator.done).toEqual(false);
      generator.next();
      expect(present(generator)).toMatchSnapshot();
      expect(generator.done).toEqual(false);
      generator.next();
      expect(present(generator)).toMatchSnapshot();
      expect(generator.done).toEqual(false);
      generator.next();
      expect(present(generator)).toMatchSnapshot();
      expect(generator.done).toEqual(true);
    });

    it('generates by one for given batch from end', () => {
      const operations = [add, subtract];
      const generator = new CombinationGenerator(
        allNumbersByOneGenerator(3),
        operations,
        {
          batchNumber: 3,
          batchLength: 3,
        },
      );

      expect(present(generator)).toMatchSnapshot();
      expect(generator.done).toEqual(false);
      generator.next();
      expect(present(generator)).toMatchSnapshot();
      expect(generator.done).toEqual(true);
    });
  });
});
