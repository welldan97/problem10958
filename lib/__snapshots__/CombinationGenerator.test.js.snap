// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CombinationGenerator in batches generates by one for given batch from beginning 1`] = `
Object {
  "value": Object {
    "numbers": Array [
      1,
    ],
    "operations": Array [],
    "priorities": Array [],
  },
}
`;

exports[`CombinationGenerator in batches generates by one for given batch from beginning 2`] = `
Object {
  "value": Object {
    "numbers": Array [
      1,
      2,
    ],
    "operations": Array [
      "add",
    ],
    "priorities": Array [
      0,
    ],
  },
}
`;

exports[`CombinationGenerator in batches generates by one for given batch from beginning 3`] = `
Object {
  "value": Object {
    "numbers": Array [
      1,
      2,
    ],
    "operations": Array [
      "subtract",
    ],
    "priorities": Array [
      0,
    ],
  },
}
`;

exports[`CombinationGenerator in batches generates by one for given batch from end 1`] = `
Object {
  "value": Object {
    "numbers": Array [
      1,
      2,
      3,
    ],
    "operations": Array [
      "subtract",
      "subtract",
    ],
    "priorities": Array [
      0,
      1,
    ],
  },
}
`;

exports[`CombinationGenerator in batches generates by one for given batch from end 2`] = `
Object {
  "value": Object {
    "numbers": Array [
      1,
      2,
      3,
    ],
    "operations": Array [
      "subtract",
      "subtract",
    ],
    "priorities": Array [
      1,
      0,
    ],
  },
}
`;

exports[`CombinationGenerator in batches generates by one for given batch from middle 1 1`] = `
Object {
  "value": Object {
    "numbers": Array [
      1,
      2,
      3,
    ],
    "operations": Array [
      "add",
      "add",
    ],
    "priorities": Array [
      0,
      1,
    ],
  },
}
`;

exports[`CombinationGenerator in batches generates by one for given batch from middle 1 2`] = `
Object {
  "value": Object {
    "numbers": Array [
      1,
      2,
      3,
    ],
    "operations": Array [
      "add",
      "add",
    ],
    "priorities": Array [
      1,
      0,
    ],
  },
}
`;

exports[`CombinationGenerator in batches generates by one for given batch from middle 1 3`] = `
Object {
  "value": Object {
    "numbers": Array [
      1,
      2,
      3,
    ],
    "operations": Array [
      "add",
      "subtract",
    ],
    "priorities": Array [
      0,
      1,
    ],
  },
}
`;

exports[`CombinationGenerator in batches generates by one for given batch from middle 2 1`] = `
Object {
  "value": Object {
    "numbers": Array [
      1,
      2,
      3,
    ],
    "operations": Array [
      "add",
      "subtract",
    ],
    "priorities": Array [
      1,
      0,
    ],
  },
}
`;

exports[`CombinationGenerator in batches generates by one for given batch from middle 2 2`] = `
Object {
  "value": Object {
    "numbers": Array [
      1,
      2,
      3,
    ],
    "operations": Array [
      "subtract",
      "add",
    ],
    "priorities": Array [
      0,
      1,
    ],
  },
}
`;

exports[`CombinationGenerator in batches generates by one for given batch from middle 2 3`] = `
Object {
  "value": Object {
    "numbers": Array [
      1,
      2,
      3,
    ],
    "operations": Array [
      "subtract",
      "add",
    ],
    "priorities": Array [
      1,
      0,
    ],
  },
}
`;

exports[`CombinationGenerator in batches generates by one for given batch which starts in the middle of number iteration and ends in the middle of the next one 1`] = `
Object {
  "value": Object {
    "numbers": Array [
      1,
      2,
      3,
    ],
    "operations": Array [
      "subtract",
      "add",
    ],
    "priorities": Array [
      1,
      0,
    ],
  },
}
`;

exports[`CombinationGenerator in batches generates by one for given batch which starts in the middle of number iteration and ends in the middle of the next one 2`] = `
Object {
  "value": Object {
    "numbers": Array [
      1,
      2,
      3,
    ],
    "operations": Array [
      "subtract",
      "subtract",
    ],
    "priorities": Array [
      0,
      1,
    ],
  },
}
`;

exports[`CombinationGenerator in batches generates by one for given batch which starts in the middle of number iteration and ends in the middle of the next one 3`] = `
Object {
  "value": Object {
    "numbers": Array [
      1,
      2,
      3,
    ],
    "operations": Array [
      "subtract",
      "subtract",
    ],
    "priorities": Array [
      1,
      0,
    ],
  },
}
`;

exports[`CombinationGenerator in batches generates by one for given batch which starts in the middle of number iteration and ends in the middle of the next one 4`] = `
Object {
  "value": Object {
    "numbers": Array [
      1,
      2,
      3,
      4,
    ],
    "operations": Array [
      "add",
      "add",
      "add",
    ],
    "priorities": Array [
      0,
      1,
      2,
    ],
  },
}
`;

exports[`CombinationGenerator without batches generates by one 1`] = `
Object {
  "value": Object {
    "numbers": Array [
      1,
    ],
    "operations": Array [],
    "priorities": Array [],
  },
}
`;

exports[`CombinationGenerator without batches generates by one 2`] = `
Object {
  "value": Object {
    "numbers": Array [
      1,
      2,
    ],
    "operations": Array [
      "add",
    ],
    "priorities": Array [
      0,
    ],
  },
}
`;

exports[`CombinationGenerator without batches generates by one 3`] = `
Object {
  "value": Object {
    "numbers": Array [
      1,
      2,
    ],
    "operations": Array [
      "subtract",
    ],
    "priorities": Array [
      0,
    ],
  },
}
`;

exports[`CombinationGenerator without batches generates by one 4`] = `
Object {
  "value": Object {
    "numbers": Array [
      1,
      2,
      3,
    ],
    "operations": Array [
      "add",
      "add",
    ],
    "priorities": Array [
      0,
      1,
    ],
  },
}
`;

exports[`CombinationGenerator without batches generates by one 5`] = `
Object {
  "value": Object {
    "numbers": Array [
      1,
      2,
      3,
    ],
    "operations": Array [
      "add",
      "add",
    ],
    "priorities": Array [
      1,
      0,
    ],
  },
}
`;

exports[`CombinationGenerator without batches generates by one 6`] = `
Object {
  "value": Object {
    "numbers": Array [
      1,
      2,
      3,
    ],
    "operations": Array [
      "add",
      "subtract",
    ],
    "priorities": Array [
      0,
      1,
    ],
  },
}
`;

exports[`CombinationGenerator without batches generates by one 7`] = `
Object {
  "value": Object {
    "numbers": Array [
      1,
      2,
      3,
    ],
    "operations": Array [
      "add",
      "subtract",
    ],
    "priorities": Array [
      1,
      0,
    ],
  },
}
`;

exports[`CombinationGenerator without batches generates by one 8`] = `
Object {
  "value": Object {
    "numbers": Array [
      1,
      2,
      3,
    ],
    "operations": Array [
      "subtract",
      "add",
    ],
    "priorities": Array [
      0,
      1,
    ],
  },
}
`;

exports[`CombinationGenerator without batches generates by one 9`] = `
Object {
  "value": Object {
    "numbers": Array [
      1,
      2,
      3,
    ],
    "operations": Array [
      "subtract",
      "add",
    ],
    "priorities": Array [
      1,
      0,
    ],
  },
}
`;

exports[`CombinationGenerator without batches generates by one 10`] = `
Object {
  "value": Object {
    "numbers": Array [
      1,
      2,
      3,
    ],
    "operations": Array [
      "subtract",
      "subtract",
    ],
    "priorities": Array [
      0,
      1,
    ],
  },
}
`;

exports[`CombinationGenerator without batches generates by one 11`] = `
Object {
  "value": Object {
    "numbers": Array [
      1,
      2,
      3,
    ],
    "operations": Array [
      "subtract",
      "subtract",
    ],
    "priorities": Array [
      1,
      0,
    ],
  },
}
`;
